pkg/codegen/schema/schema.go

```diff
@@ -1539,6 +1641,152 @@ type FunctionSpec struct {
 	IsOverlay bool `json:"isOverlay,omitempty" yaml:"isOverlay,omitempty"`
 }
 
+func emptyObject(data RawMessage) (bool, error) {
+	var objectData *map[string]RawMessage
+	if err := json.Unmarshal(data, &objectData); err != nil {
+		return false, err
+	}
+
+	if objectData == nil {
+		return true, nil
+	}
+
+	return len(*objectData) == 0, nil
+}
+
+func UnmarshalFunctionSpec(funcSpec *FunctionSpec, data map[string]RawMessage) error {
```

Comment:
Why do we need to custom unmarshal the entire `FunctionSpec`? This seems really error prone.

Also, does this need to be public?

Reply:
Because I need  to control how two fields are unmarshalled: `Outputs` and `ReturnType`, the former being derived from the latter. Same goes for marshalling. However, it does not need to be public. I will change that

Reply:
Does this need to be done at the `FunctionSpec` level. I imagine it would be much easier and less error prone to do at the bind level. 

P.S. The current implementation directly calls `json.Unmarshal`, which won't work for yaml.

Reply:
> Does this need to be done at the FunctionSpec level. I imagine it would be much easier and less error prone to do at the bind level.

I tried removing the unmarshalling logic and instead only have custom marshalling for `ReturnTypeSpec`, however that doesn't work with YAML because if I put the attribute `yaml:"outputs"` on the `ReturnTypeSpec`, then it fails saying that there are duplicate `outputs` definitions because the `Outputs` fields is still there. It can be simplified when we remove `Outputs` entirely

> P.S. The current implementation directly calls json.Unmarshal, which won't work for yaml.

It works because it uses `map[string]RawMessage` as an intermediate data structure for both JSON and YAML. See unit tests I've added in `schema_test.go` for marshalling and unmarshalling `FunctionSpec`